transport: load balancing module refactor

havaker opened the pull request:
## Description

This is the third attempt to refactor this module;)

#### Replica set calculation was moved to a separate module (`transport::locator`).

Replica set calculation is done by a module originally written by @cvybhu - `transport::locator::replication_info`. `struct ReplicationData` lives in that module, and provides a set of functions that allow calculating of `SimpleStrategy` & `NetworkTopologyStrategy` replica lists.

To make load balancing fast, precomputing replica sets is required. This is done by another of @cvybhu's modules - `transport::locator::precomputed_replicas`. `precomputed_replicas::PrecomputedReplicas` is a struct that precomputes replica lists of a given strategies, and provides O(1) access to desired replica slices.

`locator::ReplicaLocator` combines the functionality of the previously mentioned modules, and provides a unified API for getting precomputed or calculated on the fly replica sets. Representing replica sets by a custom `ReplicaSet` type allowed creating an API that supported optional limiting replicas to a single data center (and allocation-free creation, getting & iteration through precomputed replica sets). 

Things left for a follow-up:
- A way to configure which strategies need replica set precomputation

#### `LoadBalancingPolicy` interface was changed.

`plan` was split to `pick` and `fallback` methods. This allows to better optimize the most common case, where only one node from the load balancing plan is needed. Changes required in query execution code were minimized by providing a lazy chaining iterator `ltransport::load_balancing::plan::Plan`. This iterator's first element is a node returned by the `LoadBalancingPolicy::pick` function, the next items come from `LoadBalancingPolicy::fallback` iterator. Falback method is called lazily - only when second+ element of the `Plan` iterator is needed.

The following methods were added:
```rust
fn on_query_success(&self, query: &QueryInfo, latency: Duration, node: NodeRef<_'>);
fn on_query_error(&self, query: &QueryInfo, latency: Duration, node: NodeRef<_'>, error: &QueryError);
```
The motivation for adding them was the ability to contain the logic of latency-aware policy inside some load balancing policy.

#### Default policy was created.

Default policy supports token and data center awareness. It also has a logic to do a data center failover.

Things left for a follow-up:
- LWT routing support

#### Latency aware policy was merged with the default one

Mechanisms used only by latency aware policy were removed (e.g. `TimestampedAverage` living in `Node`).

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

Fixes: https://github.com/scylladb/scylla-rust-driver/issues/408

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -1,25 +1,334 @@
+        routing_info
+    }
+
+    fn preffered_node_set<'a>(&'a self, cluster: &'a ClusterData) -> &'a [Arc<Node>] {
```
preffered -> prefered

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -1,25 +1,334 @@
+    }
+
+    fn preffered_node_set<'a>(&'a self, cluster: &'a ClusterData) -> &'a [Arc<Node>] {
+        if let Some(preffered_datacenter) = &self.preferred_datacenter {
```
ditto

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -1,25 +1,334 @@
+                return nodes;
+            } else {
+                tracing::warn!(
+                    "Datacenter specified as the preffered one ({}) does not exist!",
```
ditto

wprzytula commented:
I'm not familiar with the token ring, replica computation etc., but apart from that, this approach looks good.

piodul commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -1,25 +1,334 @@
-    pub fn new() -> Self {
-        Self {}
+impl LoadBalancingPolicy for DefaultPolicy {
+    fn pick<'a>(&'a self, query: &'a QueryInfo, cluster: &'a ClusterData) -> NodeRef<'a> {
```
I've got a feeling that `pick` and `fallback` methods follow a similar pattern (control flow-wise) and we could try to reduce the complexity by deduplicating them. Basically, you could think of `pick` as a function that returns the first element of `fallback` without having to allocate memory for the `FallbackPlan`. Of course, if we implemented `pick` such that it calls` fallback` then the performance benefit would be gone - however, what if we removed the need to allocate the plan with a callback?

You could implement a function `do_with_plan` (didn't think too much about the name):

```rust
// FnOnce won't be sufficient to express callback's type, because you don't have
// access to the iterator's type - you will most likely have to use a trait
trait PlanExtractor<'a> {
    type Output;
    fn extract<I>(plan: I) -> Self::Output
    where
        I: Iterator<Output = NodeRef<'a> + Send + Sync + 'a>;
}

fn do_with_plan<'a, E>(
    &'a self,
    query: &'a QueryInfo,
    cluster: &'a ClusterData,
    extractor: E,
) -> E::Output
where
    E: PlanExtractor<'a>,
{
    /* snip */

    let plan = maybe_replicas
        .chain(robined_local_nodes)
        .chain(maybe_remote_nodes)
        .unique();

    extractor.extract(plan)
}
```

I didn't really scrutinize the `pick` and `fallback` methods so I'm not sure what the performance impact of this approach will be. If computing the plan requires significantly more computations than the current implementation of `pick` (apart from avoiding the allocation), then I guess we can stay with the current approach.

It would be great if we could compare the performance of both approaches. Perhaps a benchmark would be in order? After all, one of the reasons for the refactor was to improve the performance.

cvybhu commented on examples/compare-tokens.rs:
```diff
@@ -31,14 +31,15 @@ async fn main() -> Result<()> {
         let serialized_pk = (pk,).serialized()?.into_owned();
         let t = Murmur3Partitioner::hash(prepared.compute_partition_key(&serialized_pk)?).value;
 
-        let statement_info = scylla::transport::load_balancing::Statement {
```
It would be nice to put the name change of `Statement` to `QueryInfo` in a separate commit. The way it's now it makes it much harder to get acquainted with the new interface because of all the clutter.

havaker replied:
I have the impression that the added value of extracting it into separate commit would require a disproportionate amount of work now.

cvybhu commented on scylla/tests/retries.rs:
```diff
@@ -29,22 +29,20 @@ fn init_logger() {
+        cluster: &'a ClusterData,
+    ) -> load_balancing::FallbackPlan<'a> {
+        Box::new(cluster.get_nodes_info().iter())
     }
```
Shouldn't there be the `sorted_by`?

havaker replied:
There is now.

cvybhu commented on scylla/src/transport/mod.rs:
```diff
@@ -6,6 +6,7 @@ pub mod downgrading_consistency_retry_policy;
 pub mod host_filter;
 pub mod iterator;
 pub mod load_balancing;
+mod locator;
```
This belongs to a different commit

havaker replied:
Fixed.

cvybhu commented on scylla/src/transport/locator/replication_info.rs:
```diff
@@ -0,0 +1,183 @@
+            if self.acceptable_repeats > 0 {
+                self.acceptable_repeats -= 1;
+            } else {
+                return None;
```
I agree that the logic looks really strange here. I think it was written by me a while ago it but it seems wrong.

cvybhu reviewed (commented):
Had an initial look, generally looks nice although I didn't dig into the juicy details yet.

cvybhu commented on scylla/src/transport/locator/replication_info.rs:
```diff
@@ -0,0 +1,183 @@
+
+    pub fn simple_strategy_replicas(
+        &self,
+        token: impl Borrow<Token>,
```
IIRC I originally did `impl Borrow<Token>` because it was annoying to add dereferences/references everywhere. It doesn't hurt to accept both references and values here.

cvybhu commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -1,25 +1,334 @@
+
+        // Get a list of all local alive nodes, and apply a round robin to it
+        let local_nodes = self.preffered_node_set(cluster);
+        let robined_local_nodes = Self::round_robin_nodes(local_nodes, Self::is_alive);
```
Round robin has the problem of overloads in case of node failure.
If the usual round robin order is A->B->C and A fails, B will take over all of A's requests. It would be better to try nodes in random order, then in case of A's failure the load that used to be handled by A will be shared equally among B and C.

wprzytula replied:
As we have no quick idea how to mitigate this problem yet, let's put off this problem until a follow-up.

cvybhu reviewed (commented):
Looks nice.
We got rid of the problematic policy chaining and the code is much clearer.

I like how simple planning has become, in my implementation I had a bunch of convoluted iterators that were a pain to write and read. This implementation is much more elegant.

The direction seems good, but there are still a few things that need to be solved before proceeding, like LWT routing, latency awareness and proper UP/DOWN event handling.

havaker commented on scylla/src/transport/locator/precomputed_replicas.rs:
```diff
@@ -0,0 +1,138 @@
+#[derive(Clone)]
+pub struct PrecomputedReplicas {
+    global_replicas: PrecomputedReplicasRing,
+    datacenter_replicas: HashMap<String, PrecomputedReplicasRing>,
```
In current solution we are leveraging an advantageous property of `SimpleStrategy` and datacenter-local `NetworkTopologyStrategy` - replica lists for replication factor `n` are prefixes of replica lists for replication factor `n + 1`. This enables us to only compute replica lists for maximal replication factor found in replication strategies (in `NTS`, we are interested in maximal `rf` specified for each datacenter). It is a big optimization - other solution is to compute those lists for each strategy used in cluster.
Computing replica lists for `MergedNTS` would unfortunately disallow us from using the optimization mentioned earlier. For a better picture, let's consider 2 example `NTS` strategies.
```
let A = NTS with rf = 1 in eu, rf = 2 in us
let B = NTS with rf = 2 in eu, rf = 1 in us
```
If `la` - replica list created using strategy `A`, `lb` - replica list created using strategy `B`, then then neither is a prefix of the other.

So, how would you imagine computing `MergedNTS` in an efficient manner?

avelanarius commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -0,0 +1,334 @@
+        // If no token was available (or all the replicas for that token are down), try to pick
+        // some alive local node.
+        // If there was no preffered datacenter specified, all nodes are treated as local.
+        let nodes = self.preffered_node_set(cluster);
```
```suggestion
        let nodes = self.preferred_node_set(cluster);
```

wprzytula commented:
I'm sad to see that we didn't really take advantage of grouping all default features in one Default Policy. Namely, we still pass the load balancing policy in an Arc, which imposes some overhead. Instead, I would propose:

```
struct LoadBalancing(Or)

enum Or {
    Default(DefaultLoadBalancingPolicy),
    Custom(Arc<dyn LoadBalancingPolicy>)
}

impl LoadBalancingPolicy for LoadBalancing {
   ... 
}
```

wprzytula commented on scylla/src/transport/locator/replication_info.rs:
```diff
@@ -0,0 +1,183 @@
+    unique_nodes_in_global_ring: Vec<Arc<Node>>,
+
+    pub(crate) datacenters: HashMap<String, DatacenterNodes>,
+    empty_dc_data: DatacenterNodes,
```
Where would it live? How would it be accessed? Do you imply employing `lazy_static!` (or equivalent)?

wprzytula commented on scylla/src/transport/locator/replication_info.rs:
```diff
@@ -0,0 +1,183 @@
+        };
+
+        let cur_rack: Option<&str> = next_node.rack.as_deref();
+        if self.used_racks.contains(&cur_rack) {
```
Interesting thing here is that `BTreeSet::insert()` returns a bool instead of an Option, and it does not replace an old value with the new in case one was already present.

wprzytula commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -4,5 +4,448 @@ pub mod replication_info;
+///
+/// This container can only be created by calling `ReplicaLocator::replicas_for_token`, and it
+/// can borrow precomputed replica lists living in the locator.
+pub struct ReplicaSet<'a> {
```
I've spent quite a lot of time on considering alternatives, and I think that the current names are the best. What could indeed be done is making their docstrings explaining the difference clearly.

havaker replied:
How about changing `Replicas` to `ReplicasArray`? The new name will carry information about contiguous way of storing elements.

wprzytula commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -4,5 +4,448 @@ pub mod replication_info;
+    }
+}
+
+enum ReplicaSetInner<'a> {
+    Plain(Replicas<'a>),
+
+    // Represents a set of SimpleStrategy replicas that is limited to a specified datacenter.
+    FilteredSimple {
+        replicas: Replicas<'a>,
+        datacenter: &'a str,
+    },
+
+    // Represents a set of NetworkTopologyStrategy replicas that is not limited to any specific
+    // datacenter. The set is constructed lazily, by invoking
+    // `locator.get_network_strategy_replicas()`.
+    MergedNTS {
+        datacenter_repfactors: &'a HashMap<String, usize>,
+        locator: &'a ReplicaLocator,
+        token: Token,
+    },
+}
```
I think that what the names mean here is what is done with various `Replicas` sequences in order to create the iterator:
- in Plain, there already is one consecutive sequence that is to be iterated,
- in Filtered, such one sequence will be iterated with a predicate filter,
- in Merged, many sequences will be chained (or "merged") in order to create the iterator.

We _could_ change the names into e.g.:
- Plain -> Sequence
- Filtered -> FilteredSequence
- Merged -> ChainedSequences
...but I believe the current names are better. At most, we could replace `Merged` with `Chained`.

havaker replied:
> I think that the variant names don't describe well what they are supposed to mean, especially the first two ones.
> 
> `Precomputed` `FilteredByDc` `MergedNts` - I don't have a better suggestion here

In addition to what @wprzytula wrote, `Plain` variant does not necessarily represents a precomputed replica sequence. It is because the `Replicas` container may hold:
- a slice to a precomputed sequence
- `Vec` that contains a sequence that was computed on the fly.

havaker commented:
Rebased on top of main (using @wprzytula's pull request https://github.com/havaker/scylla-rust-driver/pull/8)

havaker commented:
Removed `scylla/src/transport/load_balancing/latency_aware.rs` in `transport: load_balancing: remove policies other than the default one`.

havaker commented:
Rebased on top of main, merged https://github.com/havaker/scylla-rust-driver/pull/6 (with additional `NodeRef` import fixes)

havaker commented on scylla/src/transport/locator/token_ring.rs:
```diff
@@ -0,0 +1,173 @@
+use std::borrow::Borrow;
+
+#[derive(Clone)]
+pub struct TokenRing<ElemT> {
```
This is a user-facing struct, it's returned by `transport::locator::ReplicaLocator::ring()` method.

havaker commented on scylla/src/transport/locator/replicas.rs:
```diff
@@ -0,0 +1,69 @@
+/// This type is very similar to `Cow`, but unlike the previously mentioned it holds references in a
+/// `Owned` variant `Vec`.
+#[derive(Debug)]
+pub enum Replicas<'a> {
```
`transport::locator::replicas` is a private module. We don't need to add `(crate)`, as it won't restrict access any further.

It is possible though (@wprzytula suggestion) is to use `pub(crate)` even in private modules, to make it more explicit that these parts of our code are not a part of public API.
What do you think of using `pub(crate)` like this?

havaker commented on scylla/src/transport/locator/precomputed_replicas.rs:
```diff
@@ -0,0 +1,138 @@
+pub type Replicas = Vec<Arc<Node>>;
+
+#[derive(Clone)]
+pub struct PrecomputedReplicas {
```
I've added explanation comments together with @cvybhu . FYI `PrecomputedReplicas` is not a user-facing struct.

havaker commented on scylla/src/transport/locator/precomputed_replicas.rs:
```diff
@@ -0,0 +1,138 @@
+        keyspace_strategies: impl Iterator<Item = &'a Strategy>,
+    ) -> PrecomputedReplicas {
+        // Each ring will precompute for at least this RF
+        let min_precomputed_rep_factor: usize = 3;
```
Changed to 1.

havaker commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -4,5 +4,448 @@ pub mod replication_info;
+            }
+        }
+    }
+}
```
Tests were added to every locator module (except for `locator::replicas`).

havaker commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -4,5 +4,448 @@ pub mod replication_info;
+                    }
+
+                    nodes_to_skip -= *repfactor;
+                }
```
I'll add a documentation note. Token order should not be too important except for LWT.

havaker commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -1,25 +1,334 @@
+    }
+
+    fn is_alive(node: &NodeRef<'_>) -> bool {
+        node.is_enabled() && !node.is_down()
```
We've given up on `!node.is_down()` check for now.

havaker commented:
> I'm sad to see that we didn't really take advantage of grouping all default features in one Default Policy. Namely, we still pass the load balancing policy in an Arc, which imposes some overhead. Instead, I would propose:
> 
> ```
> struct LoadBalancing(Or)
> 
> enum Or {
>     Default(DefaultLoadBalancingPolicy),
>     Custom(Arc<dyn LoadBalancingPolicy>)
> }
> 
> impl LoadBalancingPolicy for LoadBalancing {
>    ... 
> }
> ```

I don't think that the overhead you mentioned is noticeable.

havaker commented on scylla/src/transport/locator/replicas.rs:
```diff
@@ -0,0 +1,69 @@
+    }
+}
+
+pub const EMPTY_REPLICAS: Replicas<'static> = Replicas::Borrowed(&[]);
```
Please see [my reply in another thread](https://github.com/scylladb/scylla-rust-driver/pull/612#discussion_r1102679213).

havaker commented:
v1:
- Rebased on top of main
- Prepared a commit series that is suitable for review

havaker commented:
v2:
- applied clippy suggestions
- added doc comments to the new `LoadBalancingPolicy` interface

havaker commented:
Oops, I left a https://github.com/scylladb/scylla-rust-driver/pull/612/commits/551485e5936e70195741747c1d815b63e2021737 that I intended to squash with `transport: load_balancing: remove ChildLoadBalancingPolicy trait`.

havaker commented:
v2.1:
- squashed https://github.com/scylladb/scylla-rust-driver/commit/551485e5936e70195741747c1d815b63e2021737

havaker commented:
> Please update the PR description. It mentions that some things still need to be done - the PR was marked as ready so I guess those things are ready?

Updated, I've added `Things left for a follow-up` lists.

havaker commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -0,0 +1,522 @@
+    /// locator, this function will have O(D * N) complexity, where N - number of all nodes.
+    ///
+    /// In all other cases, the complexity is O(1)
+    pub fn get(&self, index: usize) -> Option<NodeRef<'a>> {
```
I'm wondering if it makes sense for this method to exist. It has a weird complexity when `NetworkTopologyStrategy` replicas are not precomputed. It also seems strange to me to use indexing in a type which name ends with `Set`.
I can't think of a valid use case for invoking `get(3)` (except for choosing randomly).

Maybe a better idea would be to throw this method away and add a `pick_random(predicate)` that would be arguably more useful in load balancing policies creation context.

wprzytula replied:
Is this comment resolved?

piodul reviewed (commented):
(Leaving some comments for now, I will continue review tomorrow)

havaker commented:
v2.2:
- changed the order of commits, moving locator-related stuff to the beginning

wprzytula commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -939,3 +980,1000 @@ mod tests {
+        }
+    }
+    /*
+    #[cfg(test)]
```
The commented out code comes from the previous implementation of LAP. They are going to be rewritten in order to test the new impl.

wprzytula replied:
This is a good question: do we insist on testing LAP in this PR or not. I would prefer not, and then we can delete the commented out code.

piodul replied:
To be completely honest, I would prefer to have them in this PR. I would rather not commit such large pieces of code as the latency aware policy without tests.

havaker commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -4,3 +4,465 @@ mod replication_info;
+
+    /// Returns a set of nodes that are considered to be replicas for a given token and strategy.
+    /// If the `datacenter` parameter is set, the returned `ReplicaSet` is limited only to replicas
+    /// from that datacenter. If a specified datacenter name does not correspond to a valid
+    /// datacenter, an empty set will be returned.
```
I believe it is. If a datacenter is invalid, the SimpleStrategy fallback will return an empty set.

That said, after detecting that the dc is invalid the code should just return an empty set, not doing any fallbacks.

havaker commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -4,3 +4,465 @@ mod replication_info;
+                        )
+                        .into();
+                    } else {
+                        warn!("Datacenter ({}) does not exist! Falling back to SimpleStrategy with replication_factor = 1", dc);
```
Misconfiguration of `DefaultPolicy` - providing a `preferred_datacenter` name that does not match anything in the metadata received from the cluster.

havaker commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -4,3 +4,465 @@ mod replication_info;
+                }
+            }
+            Strategy::Other { name, .. } => {
+                warn!("Unknown strategy ({}), falling back to SimpleStrategy with replication_factor = 1", name)
```
I'll change the log level to `debug`.

havaker commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -355,3 +355,165 @@ impl<'a> TokenWithStrategy<'a> {
+            let expected_groups = [&[1u16] as &[u16], &[3, 2, 4], &[5]]
+                .into_iter()
+                .map(|nodes| nodes.into_iter().copied().collect::<HashSet<u16>>())
+                .collect::<Vec<_>>();
```
@wprzytula could you please address it?

havaker commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -355,3 +355,165 @@ impl<'a> TokenWithStrategy<'a> {
+            }
+        }
+
+        pub(crate) fn assert_proper_grouping_in_plan(
```
@wprzytula could you please address it?

havaker commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -516,4 +536,389 @@ mod tests {
+                    .group([D, E, F]) // remote nodes
+                    .build(),
+            },
+            // No keyspace implies no token awareness
```
I wasn't quite sure what to do in this case. Disabling token awareness was the easiest option. Falling back to simple strategy with `rf=1` is probably more reasonable, but I'm not sure whether it won't be able to cause uneven load distribution across replicas (overloading the primary one).

piodul replied:
To be fair, now that I looked how `RoutingInfo::keyspace` is computed, I can't think of a scenario where `keyspace` can be `None` for a prepared statement, so I guess it's fine for now.

havaker commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -0,0 +1,21 @@
+#[derive(Debug, Default)]
+pub struct DefaultPolicy {}
+
+impl DefaultPolicy {
```
I wanted to remove old policies before the interface change. Removing them all would break the assumption that the `Session` has some default policy available. So in order not to break tests I chose to introduce the `DefaultPolicy` early, and incrementally improve it further in the series.

havaker commented on scylla/src/transport/load_balancing/mod.rs:
```diff
@@ -100,166 +100,3 @@ fn slice_rotated_left<T>(slice: &[T], mid: usize) -> impl Iterator<Item = &T> +
     let end = &slice[..mid];
     begin.iter().chain(end.iter())
 }
-
```
Sorry, I try to do that in the future:/

havaker commented on examples/execution_profile.rs:
```diff
@@ -21,9 +21,7 @@ async fn main() -> Result<()> {
-        .load_balancing_policy(Arc::new(TokenAwarePolicy::new(Box::new(
-            RoundRobinPolicy::new(),
-        ))))
+        .load_balancing_policy(Arc::new(load_balancing::DefaultPolicy::default()))
```
https://github.com/scylladb/scylla-rust-driver/pull/612#discussion_r1113287802

havaker commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -0,0 +1,21 @@
+
+impl LoadBalancingPolicy for DefaultPolicy {
+    fn plan<'a>(&self, _statement: &Statement, _cluster: &'a ClusterData) -> Plan<'a> {
+        todo!();
```
I've removed it and made `plan` return something to not break tests.

havaker commented:
v3:
- fixed `ReplicaSet` len calculation, tests included
- replaced `ReplicaSet::get` with `choose_filtered` method, tests included
- limited visibility of `Replicas`, `ReplicationInfo`, `PrecomputedReplicas` with the help of `pub(crate)`
- fixed some typos that came up in the review
- added default implementations of `on_query*` methods
- added a separate commit that renames `load_balancing::Statement` to `RoutingInfo`
- canceled deletion of custom load balancing policy example
- renamed `Replicas` to `ReplicasArray`
- renamed `MergedNTS` variant of `ReplicaSetInner` to `ChainedNTS`
- refactored locator test module (reduced code duplication, every test uses the same ring mock)
- changed behavior to return an empty set on invalid datacenter in `ReplicaLocator::replicas_for_token`
- changed log level to debug for unknown strategy in `ReplicaLocator::replicas_for_token`
- avoid iterating over the entire ring when calculating NTS replicas in `ReplicationInfo` (https://github.com/scylladb/scylla-rust-driver/pull/648)
- squashed addition of `size_hint` & `nth` method into introduction of `ReplicaSetIterator`
- restored `ClusterData::get_datacenters_info`
- removed underscore from used `is_alive` argument in default policy

things yet to be done:
- separate commit for deleting the `name` method from load balancing trait
- applying derive `Debug` in a more granular way (splitting into separate commits)
- `ClusterData::new_sync`
- using `ReplicaSet::choose_filtered` in default policy code

havaker commented on scylla/src/transport/locator/replicas.rs:
```diff
@@ -0,0 +1,69 @@
+            Replicas::Borrowed(slice) => Either::Left(slice.iter()),
+            Replicas::Owned(vec) => Either::Right(vec.into_iter()),
+        }
+    }
```
This method wasn't used, so I removed it.

havaker commented on scylla/src/transport/iterator.rs:
```diff
@@ -536,6 +537,9 @@ where
                     let _ = self.metrics.log_query_latency(elapsed.as_millis() as u64);
                     self.log_attempt_success();
                     self.log_query_success();
+                    self.execution_profile
```
I've added a separate commit that fixes this issue, as it was present before any changes that I had made. https://github.com/scylladb/scylla-rust-driver/pull/612/commits/b7c502e3ecc14166bd16fb32a8117bfb17cd3667

havaker commented on scylla/src/transport/cluster.rs:
```diff
@@ -371,15 +365,9 @@ impl ClusterData {
 
-    /// Access datacenter details collected by the driver
-    /// Returned `HashMap` is indexed by names of datacenters
-    pub fn get_datacenters_info(&self) -> &HashMap<String, Datacenter> {
-        &self.datacenters
-    }
```
> If so, maybe we could change this function to construct a new hashmap and return it by value, then deprecate it and recommend some other way in the deprecation notice?

Done as you proposed.

havaker commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -355,3 +355,165 @@ impl<'a> TokenWithStrategy<'a> {
+        }
+
+        pub(crate) fn assert_proper_grouping_in_plan(
+            got: impl Iterator<Item = u16>,
+            expected_groups: impl Iterator<Item = HashSet<u16>>,
```
@wprzytula

havaker commented on scylla/src/transport/load_balancing/default.rs:
```diff
@@ -516,4 +536,389 @@ mod tests {
+                //                                      r2  r1  r1  r1  r2  r1  r1
+                expected_groups: ExpectedGroupsBuilder::new()
+                    .group([A, C, G]) // pick + fallback local replicas
+                    .group([F, D, E]) // remote replicas
```
I think that can be resolved after our private conversation.

havaker commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -448,4 +448,54 @@ impl<'a> Iterator for ReplicaSetIterator<'a> {
+
+                self.next()
+            }
+            _ => {
```
I suggest doing this as a follow-up.

havaker commented on scylla/src/transport/load_balancing/mod.rs:
```diff
@@ -25,13 +33,32 @@ pub struct Statement<'a> {
-    /// Returns name of load balancing policy
-    fn name(&self) -> String;
+    /// Returns all contact-appropriate nodes for a given query.
+    fn fallback<'a>(&'a self, query: &'a QueryInfo, cluster: &'a ClusterData) -> FallbackPlan<'a>;
```
I'm sorry, but I'm in favor of not making such gradual changes in this pull request development state. I feel it would add a lot of time and the benefit would be rather small.

piodul commented:
> > Nit: Changing ClusterData::new resulted in having to change many function to be async. How about keeping ClusterData::new sync, and invoking it in non-testing code with spawn_blocking? This should result in slightly less churn.
> 
> > Or - even better - we could have ClusterData::new_sync and a ClusterData::new which just calls new_sync in spawn_blocking.
> 
> Using any of these approaches requires us to clone everything that is passed to `ClusterData::new()` (as we want to send the task to possibly another thread, all data used by it must be owned). I'm afraid that cloning such a big amount of data is a too much waste to go for it.
> 
> On the other hand, the functions that you observed to be necessarily made async are mainly test helpers - I wouldn't consider that a problem.
> 
> WDYT @piodul @havaker ?

It's not a super important issue, just a nitpick. If it's difficult to fix or causes other problems then let's not do it. I just wanted to reduce the amount of unnecessary changes.

havaker commented on scylla/src/transport/cluster.rs:
```diff
@@ -331,20 +335,32 @@ impl ClusterData {
+            .collect();
+
+        let locator = tokio::task::spawn_blocking(move || {
+            ReplicaLocator::new(ring.into_iter(), keyspace_strategies.iter())
```
Marking as resolved due to https://github.com/scylladb/scylla-rust-driver/pull/612#issuecomment-1460385181

havaker commented:
v5:

- rebased on top of `main`

havaker commented:
v7:
- restored `name()` method (@wprzytula convinced me that it would be useful in OTel tracing)
- fixed issues around code samples in doc book after rebase
- added doc book documentation for the load balancing module

cvybhu commented on scylla/src/transport/locator/mod.rs:
```diff
@@ -0,0 +1,554 @@
+                        .unique_nodes_in_datacenter_ring(dc)
+                        .map(|nodes| nodes.len())
+                        .unwrap_or(0);
+                    dbg!(dc);
+                    dbg!(rf);
+                    dbg!(unique_nodes_in_dc_count);
```
Please remove the debug prints.
They spam the console.

havaker replied:
Ups, I thought that I got rid of them sometime. Will fix.

havaker commented:
v8:

- removed stray `dbg!` invocations
- clarified that load balancing policies do not influence to which nodes connections are being opened
- rephrased some parts of `docs/load-balancing`

havaker commented:
v9:
- merged latency-aware policy tests from https://github.com/havaker/scylla-rust-driver/pull/16 (@wprzytula)
- fixed doc comment of `mock_metadata_for_token_aware_tests` in `transport: locator: add test module`
- changed random number generator in `ReplicaLocator` tests to be deterministic (`transport: locator: change rng in tests to be deterministic`)

havaker commented:
v10:
- merged latency-aware docs from @wprzytula (https://github.com/havaker/scylla-rust-driver/pull/17)

cvybhu commented on scylla/Cargo.toml:
```diff
@@ -64,6 +64,7 @@ ntest = "0.8.1"
 criterion = "0.3"
 tracing-subscriber = { version = "0.3.14", features = ["env-filter"] }
 assert_matches = "1.5.0"
+rand_chacha = "0.3.1"
```
Is a separate crate really needed?
`rand::rngs::StdRng` also implements `SeedableRng` and it's supposed to be repeatable.

havaker replied:
`StdRng` doc says that:
> The algorithm is deterministic but should not be considered reproducible due to dependence on configuration and possible replacement in future library versions. For a secure reproducible generator, we recommend use of the [rand_chacha](https://crates.io/crates/rand_chacha) crate directly.

havaker commented on scylla/src/transport/locator/replication_info.rs:
```diff
@@ -8,8 +8,8 @@ use std::cmp;
 
-#[derive(Clone)]
 /// DatacenterNodes nodes holds a token ring in which all nodes belong to one datacenter.
+#[derive(Debug, Clone)]
```
Wait, where is the third one? I can only see 99e85ad34e76dcf37e99bf6b6ff568479aecdaa5 and 2534cce422a530a00651a3239dbd3f0f749889a9.

piodul replied:
I can see the following duplicates

- [transport: locator: add TokenRing struct](https://github.com/scylladb/scylla-rust-driver/pull/612/commits/29224b6ff8d4d2f6df560a06a150acca295375c6) and [transport: locator: add TokenRing struct](https://github.com/scylladb/scylla-rust-driver/pull/612/commits/109139afa34f7b90487fa28cac396cf4f400114b),
- [transport: locator: add ReplicationInfo struct](https://github.com/scylladb/scylla-rust-driver/pull/612/commits/99e85ad34e76dcf37e99bf6b6ff568479aecdaa5) and [transport: locator: add ReplicationInfo struct](https://github.com/scylladb/scylla-rust-driver/pull/612/commits/2534cce422a530a00651a3239dbd3f0f749889a9),
- [transport: locator: add PrecomputedReplicas struct](https://github.com/scylladb/scylla-rust-driver/pull/612/commits/9e247960d811aee199e7164c57734f27d8b9ee2d) and [transport: locator: add PrecomputedReplicas struct](https://github.com/scylladb/scylla-rust-driver/pull/612/commits/70750b40b808cf3776aee3c6bcc90a0c8fcae924)

havaker replied:
Oh I see now that 109139afa34f7b90487fa28cac396cf4f400114b and 70750b40b808cf3776aee3c6bcc90a0c8fcae924 have the same titles as 29224b6ff8d4d2f6df560a06a150acca295375c6 and 9e247960d811aee199e7164c57734f27d8b9ee2d. Will squash them.

havaker commented:
v11:
- squashed commits that added `Debug` derive into some previous ones which were named the same

havaker commented:
v12:
- reworded docs
- rebased on top of main (to make clippy happy)

cvybhu commented:
The new version of load balancing does the precomputation of replica sets for each vnode, which could turn out to be quite a bit of work, so I measured how much resources the precomputation consumes when compared to the previous version which didn't do the precomputation (usiing [lbbench](https://github.com/cvybhu/lbbench)):

Cluster: 3 datacenters, 8 nodes in each one

### A few SimpleStrategy keyspaces, RF <= 8:
Before: 1.8MB of memory used, 420µs to compute ClusterData
After: 2.5MB of memory used, 5ms to compute ClusterData

### A few big SimpleStrategt keyspaces, RF ~= 20
Before: 1.9MB of memory used, 558.229µs to compute ClusterData
After: 3.7MB of memory used, 14ms to compute ClusterData

### A few NetworkTopologyStrategy keyspaces RF <= 8
Before: 1.9MB of memory used, 454.11µs to compute ClusterData
After: 6.3MB of memory used, 25ms to compute ClusterData

### All of the above
Before: 2MB of memory used, 354.14µs to compute ClusterData
After: 7.8MB of memory used, 37ms to compute ClusterData[](url)

----

We use a bit more memory and compute power, but it's within reasonable limits.

---

One positive thing stemming from the new precomputation is that the number of allocations per request got reduced:
Before:
```
Inserts:
----------
allocs/req:                15.00
reallocs/req:               8.00
frees/req:                 15.00
bytes allocated/req:     2458.05
bytes reallocated/req:    269.06
bytes freed/req:         2456.80
(allocated - freed)/req:      1.25
----------
Sending 100000 selects, hold tight ..........
----------
Selects:
----------
allocs/req:                48.00
reallocs/req:               8.00
frees/req:                 48.00
bytes allocated/req:     5266.07
bytes reallocated/req:    209.00
bytes freed/req:         5266.00
(allocated - freed)/req:      0.07
```

After:
```
Inserts:
----------
allocs/req:                 6.01
reallocs/req:               6.00
frees/req:                  6.00
bytes allocated/req:      381.80
bytes reallocated/req:    173.05
bytes freed/req:          380.62
(allocated - freed)/req:      1.18
----------
Sending 100000 selects, hold tight ..........
----------
Selects:
----------
allocs/req:                39.00
reallocs/req:               6.00
frees/req:                 39.00
bytes allocated/req:     3190.15
bytes reallocated/req:    113.01
bytes freed/req:         3190.04
(allocated - freed)/req:      0.11
----------
```

mykaul commented:
Nice work!
